framework/src/problems/FEProblemBase.C

```diff
@@ -2541,11 +2559,71 @@ FEProblemBase::duplicateVariableCheck(const std::string & var_name,
     {
       const Variable & var =
           curr_sys_ptr->system().variable(curr_sys_ptr->system().variable_number(var_name));
+
+      // variable type
       if (var.type() != type)
-        mooseError(error_prefix,
-                   "Variable with name '",
+      {
+        const auto stringifyType = [](FEType t)
+        { return Moose::stringify(t.family) + " of order " + Moose::stringify(t.order); };
+
+        mooseError("Mismatching types are specified for ",
+                   error_prefix,
+                   "variable with name '",
                    var_name,
-                   "' already exists but is of a differing type!");
+                   "': '",
+                   stringifyType(var.type()),
+                   "' and '",
+                   stringifyType(type),
+                   "'");
+      }
+
+      // block-restriction
+      if (!(active_subdomains->size() == 0 && var.active_subdomains().size() == 0))
+      {
+        const auto varActiveSubdomains = var.active_subdomains();
+        std::set<SubdomainID> varSubdomainIDs;
+        if (varActiveSubdomains.size() == 0)
+        {
+          const auto subdomains = _mesh.meshSubdomains();
+          varSubdomainIDs.insert(subdomains.begin(), subdomains.end());
+        }
+        else
+          varSubdomainIDs.insert(varActiveSubdomains.begin(), varActiveSubdomains.end());
+
+        if (varSubdomainIDs != subdomainIDs)
+        {
+
+          // helper function: make a string from a set of subdomain ids
+          const auto stringifySubdomains = [this](std::set<SubdomainID> subdomainIDs)
+          {
+            std::stringstream s;
+            for (auto const i : subdomainIDs)
+            {
+              // do we need to insert a comma?
+              if (s.tellp() != 0)
+                s << ", ";
+
+              // insert subdomain name and id -or- only the id (if no name is given)
+              const auto subdomainName = _mesh.getSubdomainName(i);
+              if (subdomainName.empty())
+                s << i;
+              else
+                s << subdomainName << " (" << i << ")";
+            }
+            return s.str();
+          };
+
+          const std::string msg = "Mismatching block-restrictions are specified for " +
+                                  error_prefix + "variable with name '" + var_name + "': {" +
+                                  stringifySubdomains(varSubdomainIDs) + "} and {" +
+                                  stringifySubdomains(subdomainIDs) + "}";
+
+          if (_variables_deduplication_flags.isValueSet("ALLOW_MISMATCHING_BLOCK_RESTRICTIONS"))
+            mooseWarning(msg);
```

Comment:
why allow the mismatching block restrction? we cant add the variable twice

Reply:
Up until now, Moose was ignoring this issue. A certain number of Moose models (including some tests) work, even with this problem. Providing the `variable_deduplication_check` parameter helps to make it possible to continue using these models almost unchanged.

The default is nevertheless **without** `ALLOW_MISMATCHING_BLOCK_RESTRICTIONS` and by doing so, making Moose more strict.

If we remove `variable_deduplication_check` and always call `mooseError`, do we have to fix all the failing tests?

Reply:
you mean it was merging
```
[var]
[]
```
with this?
[var]
   block = a
[]
```

> If we remove variable_deduplication_check and always call mooseError, do we have to fix all the failing tests?

yes but if the burden is unreasonable we can help out

Reply:
Yes, this is the case I had i mind.

I'll remove the parameter.

BTW: This makes https://github.com/idaholab/moose/pull/28308 a bit more important.

Oh one other thing: In another PR you told me that per PR only one commit should be merged in the end. So I amend my changes. If this is not suitable for you (and it would be better to have several commits and merging in the end) please let me know.

I'll have a look on the now failing tests in the next days.

Reply:
>  In another PR you told me that per PR only one commit should be merged in the end

no hopefully I did not say that.
We just want clean commit histories, and avoiding merging multiple versions of gold files

1 commit is one way to guarantee the last part but you do not have to

Reply:
one thing is that this merge
```
[var]
[]
```
with this?
```
[var]
   block = a
[]
```

is generally not wrong. It's often just passing a block restriction from the command line arguments, which is often used in testing to test block restriction. it could also be merging two inputs base.i and additional_things.i which each contain parameters for`var`

On the other hand, in this failed test
```
solid_mechanics/test:gravity.block-gravity-kinetic-energy
```
we have variables added by the variables block and then re-added by the SolidMechanics block, which we don't want to allow. Especially here since they have different block restrictions

Reply:
Please allow me to ask again, as I am not clear about the consequences of this last post. It's not clear to me whether we should really always generate a mooseError now?